### ISSUE TITLE
Tesla Altered their Unique Charger ID's

### ISSUE THREAD

**Issue by aesculus** [aesculus/EVTO-App-Feedback#379]
The app uses the charger unique ID's to associate chargers with their peers and to relate chargers as waypoints etc.  Need to consider a means to accommodate these changes or come up with a different schema for identifying chargers.

Note these ID's are embedded in the charger database, the users trips, the shared trips and relationship tables throughout the app and server. Converting these ID's will be a big challenge and probably come with some issues or tradeoffs (ie wiping shared trips for the transition).

**Comment by EVGrokker** [comment]
A couple of observations:
- It seems inevitable that you're going to break backwards compatibility for shared trips. Best to face up to this sooner than later, with a smaller base of affected users.
- I think you really need to anticipate that you will need to revise the stored trip data format in the future, leading to…
- I'd strongly recommend including a version ID in the trip so that future versions of the app can know how to deal with it. This isn't necessarily the version of EVTO that created it, although you could include that as well. It's an version ID of the trip data format, which presumably would change far less frequently than the app version.
- I think you need to invest the time and energy now to develop an independent database format that does not rely on anyone else's proprietary IDs. This of course would require you to also write tools to harvest the public sources and convert them to your storage format.

I can envision that your tool would scrape lat/lon, friendly name, SC attributes, and any other attributes that you think would be useful in the future, including charger type and characteristics. This would anticipate Tesla adding a new, higher-power SC format, which has been rumored, and it would also allow you to include destination chargers, and even non-Tesla chargers, which would be useful when you release a version of EVTO for non-Tesla EVs. I could see that it would make sense to suck the NREL database into your database. And this could serve as the underpinnings to allow users to create Favorite charging stops, which could be subsequently edited. This suggests an architecture supporting both server and client-side storage.

In other words, this looks like a pain in the neck at the moment, but it's actually an opportunity in disguise.

**Comment by aesculus** [comment]
I think you summarized my thoughts on this nicely and we are on the same page.

I suspect that this needs to be addressed in the next release (1.2-2.0) barring another trivial maintenance release I don't see coming.  This means we just added another 2 weeks in my best guess.

**Comment by aesculus** [comment]
For now I found out which chargers just had new ID's and swapped the old for the new.

I will work on a more solid, and future proof, solution for the real next release.

**Comment by aesculus** [comment]
Right now I have a hack in that I have a routine that scans the old and new files looking for sites that have the same lat/lng but a different ID number. I then have a swap routine that exchanges the old number for the new number.

I am going to have to monitor this when some of these old numbers come live and then decide how to proceed.  Since they are all so close to their former locations I may let the trips have a slight error in them. Doubt it would make much of a difference from a routing perspective and the users will probably hardly notice or care they have been auto routed down the street. In fact they might like it better.

**Comment by EVGrokker** [comment]
FWIW, I recently received email from Tesla about two new superchargers in the Seattle area (Sequim and Arlington), and they're both in the EVTO database.

**Comment by aesculus** [comment]
Once I got the routine down it's not too bad. But I had probably better decide what to do before the next major release.

**Comment by aesculus** [comment]
Right now I have a fairly workable solution.  I am inclined not to mess with it until I see what happens when they start activating the old stations with the new numbers.

**Comment by aesculus** [comment]
I am going to close this for now unless I run into a situation I cannot handle. Still waiting for them to reuse one of the IDs. That could potentially cause a small problem where routes might end up down the street from where they were originally routed to.
